internal/pkg/merge/merger.go

```diff
@@ -0,0 +1,225 @@
+package merge
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/protobom/protobom/pkg/sbom"
+)
+
+type (
+	Merger[T any] interface {
+		Merge(*T) (*T,error)
+	}
+
+	SliceMerger[T any] interface {
+		Merge([]*T) ([]*T,error)
+	}
+
+	MergerBase[T any] struct {
+		Base *T
+	}
+
+	SliceMergerBase[T any] struct {
+		Base []*T
+	}
+)
+
+func (m *SliceMergerBase[T]) Merge(t []*T) ([]*T,error) {
+	switch any(t).(type) {
+	case []*sbom.Person:
+		pl1 := any(m.Base).([]*sbom.Person)
+		pl2 := any(t).([]*sbom.Person)
+
+		pl1 = append(pl1, pl2...)
+		var dedupedList []*sbom.Person
+
+		for _, person := range pl1 {
+			skipAdd := false
+			for _, existingPerson := range dedupedList {
+				if existingPerson.Email == person.Email {
+					existingPerson, _ = NewMerger(existingPerson).Merge(person)
+					skipAdd = true
+					break
+				}
+			}
+
+			if !skipAdd {
+				dedupedList = append(dedupedList, person)
+			}
+		}
+		return any(dedupedList).([]*T), nil
+	case []*sbom.Tool:
+		tl1 := any(m.Base).([]*sbom.Tool)
+		tl2 := any(t).([]*sbom.Tool)
+
+		tl1 = append(tl1, tl2...)
+		var dedupedList []*sbom.Tool
+
+		for _, tool := range tl1 {
+			skipAdd := false
+			for _, existingTool := range dedupedList {
+				if tool.Name == existingTool.Name && tool.Version == existingTool.Version {
+					skipAdd = true
+					existingTool, _ = NewMerger(existingTool).Merge(tool)
+					break
+				}
+			}
+
+			if !skipAdd {
+				dedupedList = append(dedupedList, tool)
+			}
+		}
+
+		return any(dedupedList).([]*T), nil
+	case []*sbom.DocumentType:
+		dtl1 := any(m.Base).([]*sbom.DocumentType)
+		dtl2 := any(t).([]*sbom.DocumentType)
+
+		dtl1 = append(dtl1, dtl2...)
+		var dedupedList []*sbom.DocumentType
+
+		for _, documentType := range dtl1 {
+			skipAdd := false
+			for _, existingDocumentType := range dedupedList {
+				if *documentType.Name == *existingDocumentType.Name {
+					existingDocumentType, _ = NewMerger(existingDocumentType).Merge(documentType)
+					skipAdd = true
+					break
+				}
+			}
+
+			if !skipAdd {
+				dedupedList = append(dedupedList, documentType)
+			}
+		}
+
+		return any(dedupedList).([]*T), nil
+	default:
+		return nil, fmt.Errorf("Error merging type %s", reflect.TypeOf(t))
+	}
+}
+
+func (m *MergerBase[T]) Merge(t *T) (*T,error) {
+	switch any(t).(type) {
+	case *sbom.Metadata:
+		m1 := any(m.Base).(*sbom.Metadata)
+		m2 := any(t).(*sbom.Metadata)
+
+		if m1.Comment == "" && m2.Comment != "" {
+			m1.Comment = m2.Comment
+		}
+
+		if m1.Id == "" && m2.Id != "" {
+			m1.Id = m2.Id
+		}
+
+		if m1.Name == "" && m2.Name != "" {
+			m1.Name = m2.Name
+		}
+
+		if m1.Version == "" && m2.Version != "" {
+			m1.Version = m2.Version
+		}
+
+		if m1.Date == nil && m2.Date != nil {
+			m1.Date = m2.Date
+		}
+
+		m1.Tools, _ = NewSliceMerger(m1.Tools).Merge(m2.Tools)
+		m1.Authors, _ = NewSliceMerger(m1.Authors).Merge(m2.Authors)
+		m1.DocumentTypes, _ = NewSliceMerger(m1.DocumentTypes).Merge(m2.DocumentTypes)
```

Comment:
I assume stuff like this is why it's in draft, just a reminder that we need to check the errors